README.md
# GDS metrics for Python apps [![Build Status](https://travis-ci.org/alphagov/gds_metrics_python.svg?branch=master)](https://travis-ci.org/alphagov/gds_metrics_python)
[![Coverage Status](https://coveralls.io/repos/github/alphagov/gds_metrics_python/badge.svg?branch=master)](https://coveralls.io/github/alphagov/gds_metrics_python?branch=master)

This project is currently maintained by the [Digital Marketplace team](https://github.com/orgs/alphagov/teams/digitalmarketplace).

GDS Metrics are in Alpha and these instructions are subject to change.

GDS Python metrics enables your [Python Flask][] web app to export performance data to [Prometheus][], you can add metrics to your app using this [GDS Python Package][].

This package is a thin wrapper around [the Python Prometheus client][] that:

* protects your `/metrics` endpoint with basic HTTP authentication for apps deployed to GOV.UK PaaS
* exposes standard metrics to be used in Grafana
```http_server_request_duration_seconds_bucket, http_server_request_duration_seconds_count, http_server_request_duration_seconds_sum, http_server_exceptions_total```

Once you’ve added this package, metrics data is served from your app's metrics endpoint and is scraped by Prometheus. This data can be turned into performance dashboards using [Grafana][].

You can read more about the Reliability Engineering monitoring solution [here][].

## Before using GDS metrics

Before using GDS metrics you should have:

* created a [Python Flask][] app running on a [Gunicorn][] HTTP Server
* deployed it to [GOV.UK Platform as a Service (PaaS)][]

## How to install metrics for Flask apps served on a Gunicorn server

To use GDS metrics you must:

1. Add the [latest version of the package][] to your `requirements.txt`, for example:

    `gds-metrics==0.2.3`

2. Run the following command to install the package:

    `pip install -r requirements.txt`

3. In your application file, usually `application.py`, create the GDSMetrics object and pass the Flask app object to the `init_app` function:

    ```python
    ...
    from gds_metrics import GDSMetrics

    app = Flask(__name__)

    metrics = GDSMetrics()
    metrics.init_app(app)
    ...
    ```

    You want to run `metrics.init_app` before any other `init_app` functions and as early in your request processing as possible to get the most accurate and reliable response timings.

4. Add/Update your Gunicorn config to import `child_exit` from the library:

    ```python
    from gds_metrics.gunicorn import child_exit
    ```

    More information about [Prometheus Gunicorn setup][].

5. Restart your server by running:

    `gunicorn -c <config file>.py <application file>:<app variable>`

    For example -

    `gunicorn -c gunicorn_config.py application:app`

6. Visit any page of your app (for example [the index page][]) to generate some site traffic

7. Visit the metrics endpoint at `/metrics` to check if the package was set up correctly. If it's set up correctly, you will see a page containing some metrics (for example `http_server_request_duration_seconds_bucket`).

## Running on GOV.UK Platform as a Service (PaaS)

When running on PaaS, citizens won’t see your metrics in production as this endpoint is automatically protected with authentication.

You can also read the official Cloud Foundry guide which has detailed information on [deploying Python apps][].

## Optional configuration

You can change the path for serving metrics (by default `/metrics`) by setting the `PROMETHEUS_METRICS_PATH` [environment variable][].

If you are running `blue-green` deployments through a cf plugin like [autopilot][] you should disable basic auth on the `\metrics` endpoint and use [IP whitelisting][] by setting the `METRICS_BASIC_AUTH` [environment variable][] to `false`. This will minimise gaps in metrics during deployment.

## How to setup extended metrics

While common metrics are recorded by default, you can also:

* record your own metrics such as how many users are signed up for your service, or how many emails it's sent
* use the Prometheus interface to set your own metrics as the metrics Python package is built on top of [the Python Prometheus client][]

You can read more about the different types of metrics available in the [Prometheus documentation][].

## Contributing

We welcome contributions. We'd appreciate it if you write tests with your changes and document them where appropriate, this will help us review them quickly.

## Licence

This project is licensed under the [MIT License][].

[Prometheus]: https://prometheus.io/
[GDS Python package]: https://pypi.org/project/gds-metrics/
[the Python Prometheus client]: https://pypi.python.org/pypi/prometheus_client
[Grafana]: https://grafana.com/
[here]: https://reliability-engineering.cloudapps.digital/#reliability-engineering
[Gunicorn]: http://gunicorn.org/
[Prometheus Gunicorn setup]: https://github.com/prometheus/client_python#multiprocess-mode-gunicorn
[Python Flask]: http://flask.pocoo.org/
[GOV.UK Platform as a Service (PaaS)]: https://www.cloud.service.gov.uk/
[latest version of the package]: https://pypi.org/project/gds-metrics/
[the index page]: http://localhost:5000/
[PaaS]: https://www.cloud.service.gov.uk/
[deploy a basic Python app]: https://docs.cloud.service.gov.uk/#deploy-a-django-app
[deploying Python apps]: https://docs.cloudfoundry.org/buildpacks/python/index.html
[environment variable]: https://docs.cloud.service.gov.uk/#environment-variables
[Prometheus documentation]: https://prometheus.io/docs/concepts/metric_types/
[MIT License]: https://github.com/alphagov/gds_metrics_python/blob/master/LICENSE
[IP whitelisting]: https://reliability-engineering.cloudapps.digital/manuals/monitor-paas-app-with-prometheus.html#ip-whitelist-your-app-metrics-endpoint


CHANGELOG.md
## 0.2.4

### Bug fixes

* You can now use versions of prometheus-client newer than v0.4.0.

## 0.2.3

### Bug fixes

* You can no longer use versions of prometheus-client newer than v0.4.0, as they do not currently work with this package

## 0.2.2
* You can now use versions of prometheus-client newer than v0.2.0 (thanks @leohemsted)

## 0.2.1
* Fixes a bug with uncaught AttributeErrors if the `before_request` function did not run before `teardown_request`

## 0.2.0
* Makes basic auth optional

## 0.1.1
* Disabled caching for the metrics endpoint.
* Set the `prometheus_multiproc_dir` before importing prometheus client.
* Fixed duplicate metrics.
* Change the metrics path label to use the request route pattern rather than the actual path to cut down on the number of timeseries returned.
* Move metrics to separate file.
* Add authentication for metrics endpoint when deployed onto PaaS.
* Compress metrics response with gzip if provided in header.
* Labels without matching rule respond with `No endpoint`.
* Expose Counter, Gauge, Summary, Histogram to users without exposing the Prometheus client library.
* Add a `child_exit` hook for gunicorn multiprocessing support.
* Added flake8 support and tests for the code.
* Added a Makefile to make it easier to run tests and publish the library to PyPI.
* Updated the README.

## 0.1.0

### Changed
* First version of gds_metrics_python wrapper.
* Added Counter and Histogram stats based on metrics delivered by [GDS metrics Ruby][].
```http_server_request_duration_seconds_bucket, http_server_request_duration_seconds_count, http_server_request_duration_seconds_sum, http_server_exceptions_total```

[GDS metrics Python]: https://github.com/alphagov/gds_metrics_python


tests/conftest.py
import os
import pytest
import time
from flask import Flask

from gds_metrics import GDSMetrics


FAKE_APP_ID = "123"
SLOW_REQUEST_DURATION = 0.1


@pytest.fixture
def app(mocker):
    flask_app = Flask(__name__)
    ctx = flask_app.app_context()
    ctx.push()

    mocker.patch.dict(
        os.environ,
        {'VCAP_APPLICATION': '{"application_id": "' + FAKE_APP_ID + '"}'}
    )
    mocker.patch('prometheus_client.generate_latest', return_value='Prometheus metrics response')

    @flask_app.route('/')
    def index():
        return 'index'

    @flask_app.route('/slow_request')
    def slow_request():
        time.sleep(SLOW_REQUEST_DURATION)
        return 'slow request'

    @flask_app.route('/exception')
    def exception():
        raise Exception()

    yield flask_app

    ctx.pop()


@pytest.fixture(scope='function')
def client(app):
    with app.test_request_context(), app.test_client() as client:
        metrics = GDSMetrics()
        metrics.init_app(app)

        yield client


@pytest.fixture(scope='function')
def client_without_basic_auth(app, mocker):
    with app.test_request_context(), app.test_client() as client:
        mocker.patch.dict(
            os.environ,
            {
                'VCAP_APPLICATION': '{"application_id": "' + FAKE_APP_ID + '"}',
                'METRICS_BASIC_AUTH': 'false'
            }
        )

        metrics = GDSMetrics()
        metrics.init_app(app)

        yield client


@pytest.fixture(scope='function')
def client_without_env_app_id(app, mocker):
    with app.test_request_context(), app.test_client() as client:
        mocker.patch.dict(os.environ, {'VCAP_APPLICATION': '{}'})

        metrics = GDSMetrics()
        metrics.init_app(app)

        yield client


scripts/run_tests.sh
#!/bin/bash
#
# Run project tests
#
# NOTE: This script expects to be run from the project root with
# ./scripts/run_tests.sh

set -o pipefail

function display_result {
  RESULT=$1
  EXIT_STATUS=$2
  TEST=$3

  if [ $RESULT -ne 0 ]; then
    echo -e "\033[31m$TEST failed\033[0m"
    exit $EXIT_STATUS
  else
    echo -e "\033[32m$TEST passed\033[0m"
  fi
}

if [[ -z "$VIRTUAL_ENV" ]] && [[ -d venv ]]; then
  source ./venv/bin/activate
fi

# install tox first
pip install tox

tox -e flake8
display_result $? 1 "Code style check"

tox -e py35
display_result $? 2 "Unit tests"


tests/test_metrics.py
import pytest
from tests.conftest import FAKE_APP_ID, SLOW_REQUEST_DURATION

VALID_AUTH_HEADER = 'Authorization', 'Bearer {}'.format(FAKE_APP_ID)


@pytest.mark.parametrize('auth_header,expected_status', [
    (None, 401),
    ([VALID_AUTH_HEADER], 200),
    ([('Authorization', 'Bearer invalid-token')], 403),
])
def test_auth_header_returns_expected_response(client, auth_header, expected_status, mocker):
    response = client.get(
        '/metrics',
        headers=auth_header
    )
    assert response.status_code == expected_status


def test_no_basic_auth_env_flag_returns_200(client_without_basic_auth):
    response = client_without_basic_auth.get('/metrics')
    assert response.status_code == 200


def test_metrics_path_without_app_id_env_does_not_need_auth(client_without_env_app_id):
    response = client_without_env_app_id.get('/metrics')
    assert response.status_code == 200


def test_app_path_does_not_need_auth(client):
    response = client.get('/')
    assert response.status_code == 200


def test_compresses_response_with_gzip_header(client):
    response = client.get(
        '/metrics',
        headers=[('Accept-Encoding', 'gzip'), (VALID_AUTH_HEADER)]
    )
    assert response.headers['Content-Encoding'] == 'gzip'


def test_does_not_compress_response_without_gzip_header(client):
    response = client.get(
        '/metrics',
        headers=[VALID_AUTH_HEADER]
    )
    assert response.headers.get('Content-Encoding') is None


def test_gzip_request_header_does_not_affect_other_paths(app, client):
    response = client.get(
        '/',
        headers=[('Accept-Encoding', 'gzip'), (VALID_AUTH_HEADER)]
    )
    assert response.headers.get('Content-Encoding') is None


def test_calls_histogram_observe(client, mocker):
    histogram_labels = mocker.patch('gds_metrics.HTTP_SERVER_REQUEST_DURATION_SECONDS.labels')
    client.get(
        '/slow_request',
        headers=[(VALID_AUTH_HEADER)]
    )
    assert histogram_labels.called
    observe_mock = histogram_labels.return_value.observe
    assert observe_mock.called
    # cannot define exact duration so expect at least the specified duration for the slow_request
    assert observe_mock.call_args_list[0][0][0] > SLOW_REQUEST_DURATION and \
        observe_mock.call_args_list[0][0][0] < SLOW_REQUEST_DURATION + 0.05


def test_requests_increases_request_counter(client, mocker):
    counter_labels = mocker.patch(
        'gds_metrics.HTTP_SERVER_REQUESTS_TOTAL.labels')
    client.get(
        '/',
        headers=[(VALID_AUTH_HEADER)]
    )
    assert counter_labels.called
    assert counter_labels.return_value.inc.called


def test_exception_increases_exception_counter(client, mocker):
    counter_labels = mocker.patch(
        'gds_metrics.HTTP_SERVER_EXCEPTIONS_TOTAL.labels')

    response = client.get(
        '/exception',
        headers=[(VALID_AUTH_HEADER)]
    )
    assert response.status_code == 500
    assert counter_labels.called
    assert counter_labels.return_value.inc.called


gds_metrics/__init__.py
import gzip
import hmac
import json
import os
from time import monotonic

from flask import abort, g, request, Response
from flask.signals import got_request_exception, request_finished

# set multiprocess temp directory before we import prometheus_client
os.environ.setdefault('prometheus_multiproc_dir', '/tmp') # noqa

import prometheus_client
from prometheus_client import multiprocess, CollectorRegistry

from .metrics import ( # noqa proxy metric types imports
    Counter,
    Gauge,
    Summary,
    Histogram,
    HTTP_SERVER_EXCEPTIONS_TOTAL,
    HTTP_SERVER_REQUEST_DURATION_SECONDS,
    HTTP_SERVER_REQUESTS_TOTAL,
)


class GDSMetrics(object):

    def __init__(self):
        self.metrics_path = os.environ.get('PROMETHEUS_METRICS_PATH', '/metrics')
        if os.environ.get("METRICS_BASIC_AUTH", "true") == "true":
            self.auth_token = json.loads(os.environ.get("VCAP_APPLICATION", "{}")).get("application_id")
        else:
            self.auth_token = False

        self.registry = CollectorRegistry()
        multiprocess.MultiProcessCollector(self.registry)

    def init_app(self, app):
        app.add_url_rule(self.metrics_path, 'metrics', self.metrics_endpoint)

        app.before_request(self.before_request)
        request_finished.connect(self.teardown_request, sender=app)
        got_request_exception.connect(self.handle_exception, sender=app)

    def metrics_endpoint(self):
        if self.auth_token:
            auth_header = request.headers.get('Authorization', '')
            if not auth_header:
                abort(401)
            elif not hmac.compare_digest(auth_header, 'Bearer {}'.format(self.auth_token)):
                abort(403)

        response = Response(
            prometheus_client.generate_latest(self.registry),
            mimetype='text/plain; version=0.0.4; charset=utf-8',
            headers={
                'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate',
            }
        )

        accept_encoding = request.headers.get('Accept-Encoding', '')

        if 'gzip' not in accept_encoding.lower():
            return response

        response.data = gzip.compress(response.data)
        response.headers['Content-Encoding'] = 'gzip'
        response.headers['Content-Length'] = len(response.data)
        response.headers['Vary'] = 'Accept-Encoding'

        return response

    def before_request(self, *args, **kwargs):
        g._gds_metrics_start_time = monotonic()

    def teardown_request(self, sender, response, *args, **kwargs):
        if hasattr(g, "_gds_metrics_start_time"):
            # There is not a guarantee that `g._gds_metrics_start_time` will exist
            # as our `before_request` function will not run if the flask app has
            # another `before_request` that runs before ours and throws an exception
            resp_time = monotonic() - g._gds_metrics_start_time
            HTTP_SERVER_REQUEST_DURATION_SECONDS.labels(
                request.method,
                request.host,
                request.url_rule.rule if request.url_rule else 'No endpoint',
                response.status_code
            ).observe(resp_time)

        HTTP_SERVER_REQUESTS_TOTAL.labels(
            request.method,
            request.host,
            request.url_rule.rule if request.url_rule else 'No endpoint',
            response.status_code
        ).inc()

        return response

    def handle_exception(self, sender, exception, *args, **kwargs):
        HTTP_SERVER_EXCEPTIONS_TOTAL.labels(type(exception)).inc()
